Framework: Deprecate `uid` in favor of `clientId`

aduth opened the pull request:
Related: #7669

This pull request seeks to deprecate all references to "UID" in the codebase with equivalent "Client ID" naming. This change is being made in an effort to promote consistency on abbreviation capitalization, where, for example, previous selectors...

- `getAdjacentBlockUid`
- `getBlockRootUID`

...become:

- `getAdjacentBlockClientId`
- `getBlockRootClientId`

Observing the changes, you should see a number of other inconsistencies having been addressed here, including references to a block's `id` where `uid` is intended.

**Note:** This is a massive pull request, which is prone to becoming stale very quickly. I'd urge that this be acted upon soon to avoid issues with merge conflicts.

**Implementation notes:**

This does include deprecations, not only for the selectors, but also covering components which we expose on `wp.editor` which accept some prop referencing a "uid". This is achieved through a new (and temporary) `withDeprecatedUniqueId` higher-order component.

**Testing instructions:**

Ensure all forms of automated tests pass.

Verify there are no lingering references to UID in the entire codebase, aside from those necessary for temporary compatibility until total removal.

Ensure that deprecations work effectively. Here's a test block as an example, whose `edit` function depends on both the `id` prop being passed and the `BlockTitle` component accepting a `uid` prop:

```js
wp.blocks.registerBlockType( 'aduth/test-uid', {
	title: 'Test UID Block',

    category: 'common',
    
    edit: function( props ) {
        return wp.element.createElement( wp.editor.BlockTitle, { uid: props.id } );
    },
    
    save: function() {
        return null;
    }
} );
```

youknowriad reviewed (commented):
I personally prefer the concisness of `uid` even if `clientId` is more descriptive but noting that `client` can have several meanings :)

Anyway, I don't care too much and this PR is very hard to review/upgrade. I kind of trust our tests for it, so I'm approving to avoid staling PR. (I took a look but it's not easy to validate all the implications)

gziolo commented:
![screen shot 2018-07-17 at 12 37 38](https://user-images.githubusercontent.com/699132/42812526-3eed2932-89be-11e8-84d3-3164c0b31b56.png)

gziolo commented on editor/components/with-deprecated-unique-id/index.js:
```diff
@@ -0,0 +1,60 @@
+			`The \`${ prop }\` prop`,
+		] ).join( ' ' );
+
+		deprecated( message, {
```
This is a really cool usage of `createHigherOrderComponent`   👍

gziolo reviewed (approved):
I also skimmed those changes. We have deprecations for both updated props in case of components and for updated selectors. We should still include some notes in the deprecation document.

aduth committed:
Framework: Deprecate `uid` in favor of `clientId`

aduth commented:
Added a broad note to the `deprecated.md` document:

>- All references to a block's `uid` have been replaced with equivalent props and selectors for `clientId`.

Rebased and did another pass of instances of `uid` in the codebase. End-to-end tests were [uncooperative locally](https://wordpress.slack.com/archives/C02QB2JS7/p1531839645000092), but hoping Travis takes more kindly to them.
